Added functionality to change the syslog server's ip address at runtime

dcorbeil opened the pull request:
One of my goal was to set the syslog network backend's IP address at runtime. This is the PR that accomplishes that task as it wasn't already supported in Zephyr. What was supported is the ability to set the IP address at compile time through `CONFIG_LOG_BACKEND_NET_SERVER`.

I'm submitting a PR here as I intend to also submit that PR upstream and I'd like another level of review as this will be my first submission to the Zephyr Project.

bradeyh commented on include/zephyr/logging/log_backend_ble.h:
```diff
@@ -30,7 +30,7 @@ typedef void (*logger_backend_ble_hook)(bool backend_status, void *ctx);
  * @brief Allows application to add a hook for the status of the BLE
  *        logger backend.
- * @details The BLE logger backend is enabled or disabled auomatically by
+ * @details The BLE logger backend is enabled or disabled automatically by
```
Please fix: Should probably omit this change, since it's unrelated.

dcorbeil replied:
Will do!

bradeyh commented on include/zephyr/logging/log_backend_net.h:
```diff
@@ -0,0 +1,27 @@
+
+#include <stdbool.h>
+
+/**
+ * @brief Allows user to set a server IP address at runtime
+ *
+ * @details This function allows the user to set an IP address at runtime before the syslog network
+ *          backend is initialized since the string address gets parsed in log_backend_api::init().
+ *          For this to work, CONFIG_LOG_BACKEND_NET_SERVER_RUNTIME needs to be enabled and is
+ *          mutually exclusive with CONFIG_LOG_BACKEND_NET_SERVER_STATIC.
+ *
+ * @param addr String that contains the IP address.
+ *
+ * @return True if parsing could be done, false otherwise.
+ */
+
+bool log_backend_net_set_addr(const char *addr);
```
Question: Should this be guarded by `#if IS_ENABLED(CONFIG_LOG_BACKEND_NET_SERVER_RUNTIME)`?

dcorbeil replied:
Good catch. Initially I figured that it wasn't necessary because `log_backend_net_set_addr()` is used no matter what but you're totally right that the declaration needs to be guarded.

bradeyh replied:
Ah, you're right--maybe just guarded by `CONFIG_LOG_BACKEND_NET`, then.

bradeyh commented on subsys/logging/backends/log_backend_net.c:
```diff
@@ -196,20 +197,31 @@ static int format_set(const struct log_backend *const backend, uint32_t log_type
+	bool ret = net_ipaddr_parse(addr, strlen(addr), &server_addr);
+
+	if (!ret) {
+		LOG_ERR("Cannot configure syslog server address");
```
Suggestion:
```suggestion
		LOG_ERR("Cannot parse syslog server address");
```

dcorbeil replied:
Yes!

bradeyh reviewed (approved):
Looks good to me, although you should probably leave out the comment fix in the BLE backend.

Suggestion: Add `log backend net` shell command to set the IP address.

Commit:
logging: runtime setting of syslog server ip address

Added functionality to change the syslog server's
ip address at runtime as well as sample for
syslog network backend

Signed-off-by: David Corbeil <david.corbeil@dynonavionics.com>

dcorbeil commented:
> Looks good to me, although you should probably leave out the comment fix in the BLE backend.
> 
> Suggestion: Add `log backend net` shell command to set the IP address.

Oops missed that one. I'm not used to seeing requests here. 

I think I will hold off on that one. Although it is a very good idea.

bryceschober commented on subsys/logging/backends/log_backend_net.c:
```diff
@@ -196,20 +197,31 @@ static int format_set(const struct log_backend *const backend, uint32_t log_type
 	return 0;
 }
 
+bool log_backend_net_set_addr(const char *addr)
```
Please fix: Shouldn't this function's definition be excluded without `LOG_BACKEND_NET_SERVER_RUNTIME`?

dcorbeil replied:
Is that comment double or triple negative?

Banter aside, probably not since it's used no matter what. As you can [see](https://github.com/DynonAvionics/zephyr/pull/3/files#diff-014012f3e0595f1ee6b9a9c8e46a5525835d1e2e13cc61c780ea9fbe1a4bb399R219) it is called from `init_net()` when `LOG_BACKEND_NET_SERVER_RUNTIME` is _not_ defined (aka when `LOG_BACKEND_NET_SERVER_STATIC` _is_ defined)

bryceschober replied:
Ugh, sorry for the brain-dead reading.

dcorbeil replied:
No worries it happens to all of us
